File: sort/mergeSort/mergeSort.py
# 합병정렬 파이썬으로 구현해보기 

def merge_sort(lists):
  if len(lists) == 0:
    return []
  if len(lists) == 1:
    return lists
  else:
    if len(lists) == 2:
      left = lists[:1]
      right = lists[1:]
    else:
      l = (len(lists)//2) + 1
      left = lists[:l]
      right = lists[l:]
    left_list = merge_sort(left) 
    right_list = merge_sort(right)

    result = []

    while True:

      if len(left_list) == 0 :
        result += right_list
        break
      elif len(right_list) == 0:
        result += left_list
        break

      if left_list[0]>right_list[0] :
        result.append(right_list[0])
        right_list.pop(0)
      elif left_list[0]<right_list[0] :
        result.append(left_list[0])
        left_list.pop(0)
      else:
        result.append(left_list[0])
        result.append(right_list[0])
        left_list.pop(0)
        right_list.pop(0)

    return result 

File: sort/mergeSort/mergeSort.js
// 재귀함수로 합병정렬 구현하기(JS)

const mergeSort = (l) => {
    if (l.length <= 1) return l;
    else {
        const mid = Math.floor(l.length / 2);
        const left = l.slice(0, mid);
        const right = l.slice(mid);
        const sortedLeft = mergeSort(left);
        const sortedRight = mergeSort(right);
        let result = []

        while (true) {
            if (sortedLeft[0] > sortedRight[0]) {
                result.push(sortedRight[0])
                sortedRight.shift()
            } else {
                result.push(sortedLeft[0])
                sortedLeft.shift()
            }

            if (sortedLeft.length === 0) {
                result = result.concat(sortedRight)
                break
            } else if (sortedRight.length === 0) {
                result = result.concat(sortedLeft)
                break
            }
        }
        return result;
    }
}

// 구현할 때 특이사항 
// 1) concat은 합친 배열을 리턴할뿐, 합쳐주지 않는다. 따라서 합친 것을 가져다 쓰려며 변수로 담아서 써야한다!
